In [1]:
import pandas as pd
import numpy as np

## LECTURA DEL ARCHIVO

Para leer el archivo se usará la librería pandas y su función read_csv, es necesario una línea de código.

In [2]:
def leer_archivo(nombre_archivo):
    return pd.read_csv(nombre_archivo,sep=' ',
                                 names=['ciudad_A','','ciudad_B',' ','distancia'],
                                 usecols=['ciudad_A','ciudad_B','distancia'],header=None)

In [3]:
ciudades_distancia = leer_archivo('rutas.txt')
ciudades_distancia.head()

ciudad_A    ciudad_B  distancia
0  AlphaCentauri     Snowdin         66
1  AlphaCentauri       Tambi         28
2  AlphaCentauri      Faerun         60
3  AlphaCentauri     Norrath         34
4  AlphaCentauri  Straylight         34

## HALLAR DISTANCIA MÍNIMA

Para hallar la distancia mínima para recorrer todas las ciudades se realizará lo siguiente:

1. Elegir una ciudad de origen y agregarla en la ruta.
2. Buscar la distancia mínima de esa ciudad con otra.
3. Sumar la distancia mínima al recorrido total y agregar la otra ciudad a la ruta.
4. Eliminar los registros de distancias de la ciudad de origen. (debido a que ya fue considerada)
5. Aginar a la otra ciudad como cuidad de origen e iterar desde el paso 2.
5. Finalizar cuando se haya considerado todas las ciudades.

Como elegir solo una ciudad de origen no asegura la solución óptima, este proceso se iterará considerando a cada ciudad como origen.

Se selecciona todas las ciudades (valores únicos)

In [4]:
ciudades = list(set(list(ciudades_distancia['ciudad_A'].unique())+list(ciudades_distancia['ciudad_B'].unique())))
print(ciudades)

['Norrath', 'Straylight', 'Snowdin', 'Tambi', 'Arbre', 'Tristram', 'Faerun', 'AlphaCentauri']


In [5]:
def hallar_ruta(df,ciudades):
    rutas_resultantes = [] # Se almacenará las rutas por cada iteración de ciudad de origen
    distancias_resultantes = [] # Se almacenará las distancias totales por cada iteración de ciudad de origen
    
    # Bucle por cada ciudad
    for i in range(len(ciudades)):
        nodo = ciudades[i] # Ciudad de origen
        ruta = [nodo] # Se agrega a la ruta
        distancia = 0 # Se define la variable para distancia
        df_aux = df.copy() # Se realiza una copia de la daya
        # El bucle de ejecuta mientras haya registros en df_aux
        while df_aux.shape[0]>0:
            # Se busca la siguiente ciudad con distancia mínima (solo se toma uno en caso de que exista varias opciones)
            Distancia_min = df_aux[df_aux.distancia == min(df_aux.distancia[(df_aux.ciudad_A == nodo) | (df_aux.ciudad_B == nodo)])].head(1)
            # Se borra los registros con la ciudad de origen debido a que ya está considerado
            df_aux = df_aux.drop(df_aux[(df_aux.ciudad_A == nodo) | (df_aux.ciudad_B == nodo)].index.tolist())
            # Se suma la distancia
            distancia += int(Distancia_min.distancia)
            # Se actualiza la ciudad de origen
            nodo = list(Distancia_min.ciudad_A)[0] if (nodo == list(Distancia_min.ciudad_B)[0]) else list(Distancia_min.ciudad_B)[0]
            # Se agrega a la ruta
            ruta.append(nodo)
        # Se agrega la ruta resultante
        rutas_resultantes.append(ruta)
        # Se agrega la distancia resultante
        distancias_resultantes.append(distancia)
    
    # Se halla la menor de las distancias
    Menor_distancia = min(distancias_resultantes)
    # Se halla la ruta con la menor de las distancias
    Mejor_ruta = [rutas_resultantes[i] for i in range(len(distancias_resultantes)) if Menor_distancia == distancias_resultantes[i] ]
    print("El mejor camino es " , Mejor_ruta , " Con una distancia total de " , Menor_distancia)

In [6]:
hallar_ruta(ciudades_distancia,ciudades)

El mejor camino es  [['Arbre', 'Tambi', 'Snowdin', 'Faerun', 'Straylight', 'Norrath', 'AlphaCentauri', 'Tristram']]  Con una distancia total de  141


###### Por lo tanto el resultado nos indica que la ditancia mínima es 141 y la ruta es ['Arbre', 'Tambi', 'Snowdin', 'Faerun', 'Straylight', 'Norrath', 'AlphaCentauri', 'Tristram']